src/common/src/monitor/connection.rs

```diff
@@ -0,0 +1,573 @@
+            }
+            Poll::Pending => {}
+        }
+        ret
+    }
+
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<Result<usize, Error>> {
+        let (inner, monitor) = MonitoredConnection::project_into(self);
+        let ret = inner.poll_write_vectored(cx, bufs);
+        match &ret {
+            Poll::Ready(Ok(size)) => {
+                monitor.on_write(*size);
+            }
+            Poll::Ready(Err(e)) => {
+                monitor.on_write_err(e);
+            }
+            Poll::Pending => {}
+        }
+        ret
+    }
+
+    fn is_write_vectored(&self) -> bool {
+        self.inner.is_write_vectored()
+    }
+}
+
+impl<C: Connection, M> Connection for MonitoredConnection<C, M> {
+    fn connected(&self) -> hyper::client::connect::Connected {
+        self.inner.connected()
+    }
+}
+
+#[cfg(not(madsim))]
+impl<C: Connected, M> Connected for MonitoredConnection<C, M> {
+    type ConnectInfo = C::ConnectInfo;
+
+    fn connect_info(&self) -> Self::ConnectInfo {
+        self.inner.connect_info()
+    }
+}
+
+pub trait MonitorNewConnection {
+    type ConnectionMonitor: MonitorAsyncReadWrite;
+
+    fn new_connection_monitor(&self, endpoint: String) -> Self::ConnectionMonitor;
+    fn on_err(&self, endpoint: String);
+}
+
+impl<C: Service<Uri>, M: MonitorNewConnection + Clone + 'static> Service<Uri>
+    for MonitoredConnection<C, M>
+where
+    C::Future: 'static,
+{
+    type Error = C::Error;
+    type Response = MonitoredConnection<C::Response, M::ConnectionMonitor>;
+
+    type Future = impl Future<Output = Result<Self::Response, Self::Error>> + 'static;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, uri: Uri) -> Self::Future {
+        let endpoint = format!("{:?}:{:?}", uri.host(), uri.port());
+        let monitor = self.monitor.clone();
+        self.inner
+            .call(uri)
+            .map(move |result: Result<_, _>| match result {
+                Ok(resp) => Ok(MonitoredConnection::new(
+                    resp,
+                    monitor.new_connection_monitor(endpoint),
+                )),
+                Err(e) => {
+                    monitor.on_err(endpoint);
+                    Err(e)
+                }
+            })
+    }
+}
+
+#[cfg(not(madsim))]
+impl<Con, E, C: Stream<Item = Result<Con, E>>, M: MonitorNewConnection> Stream
+    for MonitoredConnection<C, M>
+where
+    Con: Connected<ConnectInfo = TcpConnectInfo>,
+{
+    type Item = Result<MonitoredConnection<Con, M::ConnectionMonitor>, E>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let (inner, monitor) = MonitoredConnection::project_into(self);
+        inner.poll_next(cx).map(|opt| {
+            opt.map(|result| {
+                result.map(|conn| {
```

Comment:
Should we also monitor accept error here?

Reply:
It's like a try stream. On accept error the stream will be stopped, and in upper layer, they server will fail to listen to a port and panic. So I think we don't need to monitor the accept error here.